sqlengine/postgres_engine_state.go

```diff
@@ -0,0 +1,115 @@
+package sqlengine
+
+import (
+	"database/sql"
+
+	"github.com/lib/pq"
+	"github.com/pivotal-golang/lager"
+)
+
+const stateDBName = "broker_state"
+
+// passwordStorageVersion represents the current method we are using to store
+// passwords. If the way we store or encrypt passwords ever changes, this field
+// (which is stored in the database) will allow us to migrate more easily.
+const passwordStorageVersion = "1.0"
+
+func (d *PostgresEngine) openStateDB(logger lager.Logger, stateEncryptionKey string) (*postgresEngineState, error) {
+	logger = logger.Session("postgres-engine-state")
+
+	statement := "CREATE DATABASE " + stateDBName
+	logger.Debug("create-database", lager.Data{"statement": statement})
+	_, err := d.db.Exec(statement)
+	if err != nil {
+		// 42P04 means duplicate database - https://www.postgresql.org/docs/9.5/static/errcodes-appendix.html
+		if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "42P04" {
```

Comment:
We could do that, but I'd argue that's less readable.  It took me a moment or two to work out the chain of logic with the negations and the `||` there. I think it's easier to see what's going on as it is presently.